src/dcap_provider.cpp

```diff
@@ -897,52 +1009,138 @@ extern "C" quote3_error_t sgx_ql_get_quote_config(
+                    "error thrown, error code: %x: %s",
+                    error.code,
+                    error.what());
+            }
+            catch (const std::exception& error)
+            {
+                log(SGX_QL_LOG_ERROR,
+                    "Unknown exception thrown, error: %s",
+                    error.what());
+            }
+        }
+        if (!fetch_response_from_primary)
+        {
+            if (auto cache_hit = try_cache_get(secondary_base_url))
+            {
+                log(SGX_QL_LOG_INFO,
+                    "Fetching quote config from cache: '%s'.",
+                    secondary_base_url.c_str());
 
-            *pp_quote_config =
-                (sgx_ql_config_t*)(new uint8_t[cache_hit->size()]);
-            memcpy(*pp_quote_config, cache_hit->data(), cache_hit->size());
+                *pp_quote_config =
+                    (sgx_ql_config_t*)(new uint8_t[cache_hit->size()]);
+                memcpy(*pp_quote_config, cache_hit->data(), cache_hit->size());
 
-            // re-aligning the p_cert_data pointer
-            (*pp_quote_config)->p_cert_data =
-                (uint8_t*)(*pp_quote_config) + sizeof(sgx_ql_config_t);
+                // re-aligning the p_cert_data pointer
+                (*pp_quote_config)->p_cert_data =
+                    (uint8_t*)(*pp_quote_config) + sizeof(sgx_ql_config_t);
 
-            return SGX_QL_SUCCESS;
-        }
-        
-        const std::string eppid_json = build_eppid_json(*p_pck_cert_id);
-        const auto curl = curl_easy::create(cert_url, &eppid_json);
-        log(SGX_QL_LOG_INFO,
-            "Fetching quote config from remote server: '%s'.",
-            cert_url.c_str());
-        curl->set_headers(headers::default_values);
-        curl->perform();
+                return SGX_QL_SUCCESS;
+            }
 
-        // we better get TCB info and the cert chain, else we cannot provide the
-        // required data to the caller.
-        if ((get_raw_header(*curl, headers::TCB_INFO, nullptr) !=
-             SGX_PLAT_ERROR_OK) ||
-            (get_raw_header(*curl, headers::PCK_CERT_ISSUER_CHAIN, nullptr) !=
-             SGX_PLAT_ERROR_OK))
-        {
-            log(SGX_QL_LOG_ERROR, "Required HTTP headers are missing.");
-            return SGX_QL_ERROR_UNEXPECTED;
+            const std::string eppid_json = build_eppid_json(*p_pck_cert_id);
+            if (eppid_json.empty())
+            {
+                headers::default_values.insert(
+                    std::pair<std::string, std::string>("Content-Length", "0"));
+            }
+            curl = curl_easy::create(secondary_base_url, &eppid_json);
+            log(SGX_QL_LOG_INFO,
+                "Fetching quote config from remote server: '%s'.",
+                secondary_base_url.c_str());
+            curl->set_headers(headers::default_values);
+            curl->perform();
+
+            // Parse the body returned by service to extract
+            // 1. Certificate

```

Comment:
All this code is duplicate. It would be more readable - If common code is moved to a function extract_json_response.

Reply:
done